### ISSUE TITLE
Security Issues associated with Debug Library Functions debug.getinfo and debug,traceback

### ISSUE THREAD

**Issue by GINUGEORGE** [kernelsauce/turbo#354]
### Security Issues associated with Debug Library Functions debug.getinfo and debug,traceback

> WRT  https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2019-19391
> 
> **_debug.getinfo_** has a **_type confusion issue_** that leads to _**arbitrary memory write or read operations**_, because certain cases involving **_valid stack levels and > options_** are **_mishandled_**. 
> 
> NOTE: The **_LuaJIT project owner_** states that the **_debug libary is unsafe by definition_** and that this is **_not a vulnerability_**. When LuaJIT was originally developed, the expectation was that the **_entire debug library had no security guarantees_** and thus it made no sense to assign CVEs. However, not all users of later LuaJIT derivatives share this perspective.

> WRT  https://www.lua.org/pil/23.html
> 
> Unlike the other libraries, you should use the **_debug library with parsimony_**. First, some of **_its functionality_** is **_not_** exactly **_famous for performance._** Second, it **_breaks some sacred truths of the language_**, such as that you cannot access a local variable from outside the function that created it. 

> WRT  http://lua-users.org/wiki/DebugLibraryTutorial
> 
> _**6.10 – The Debug Library**_
> 
> This library provides the functionality of the debug interface to Lua programs. You should _**exert care when using this library**_. Several of its functions **_violate basic assumptions about Lua code_** (e.g., that variables local to a function cannot be accessed from outside; that userdata metatables cannot be changed by Lua code; that Lua programs do not crash) and therefore can **_compromise otherwise secure code_**. Moreover, **_some functions in this library may be slow_**.

**_Also Refer_** https://github.com/LuaJIT/LuaJIT/pull/526.

As per the above excerpts,  **_debug library functions_** should be avoided using in **_Production Environments_**. But in the **_turbo lua source code_**,  we can see that certain debug library functions such as **_debug.getinfo_** and **_debug.traceback_** have been used in the below mentioned files :

debug.traceback - https://github.com/kernelsauce/turbo/blob/master/turbo/ioloop.lua#L524

debug.getinfo - https://github.com/kernelsauce/turbo/blob/master/turbo/log.lua#L284

**_Is it safe to use these functions in the Production Environment ? Please justify the usage and share your suggestions/inputs._**

**Comment by kernelsauce** [comment]
Yes, I agree it is not optimal to use this. It is simple to remove the stack tracing that it is used for, its only for ease of debugging and does not serve any function beyond that. But given the circumstance that it is used its unlikely that it is exploitable. Correct me if I'm wrong.

**Comment by hippi777** [comment]
hi folks :)

i didnt give a feedback to this previously only cuz its a matter of ensuring that no abuse can happen, and ive thought that once upon the time i will check myself if theres anything to be afraid of or not or what should be done to play safe... for a sandbox, users must not reach the raw debug stuffs (neither on puc lua), but they can get tools that utilize anything from there behind the scenes without getting any ability to give these malicious input, or abuse output(?), while here the question is if a request (or any resulting stored data) that could reach these functions could do any harm to either the sysadmin or the rest of the users, but using the debug lib only takes extra caution, not changes the game... here, its about validating the input, and/or tracking down what functions use these, but on the 1st place i would worry about nasty requests in general, and i would test spec chars, marginal/extreme lengths, erroneous protocol usages and weird header usage, doing this either/both while reading the codes to see what data travels where, or via fuzzing, both to try to attack turbo and any application that is written on top of it, while probably the greatest attack-vector is still most possible in any custom app running on top of it :D

(sorry if i have said nothing but the obvious stuffs :D )
